tools/castle-editor/code/framedesign.pas

```diff
@@ -1776,6 +1786,114 @@ procedure TDesignFrame.CastleControlUpdate(Sender: TObject);
   end;
 end;
 
+procedure TDesignFrame.CastleControlDragOver(Sender, Source: TObject; X,
+  Y: Integer; State: TDragState; var Accept: Boolean);
+var
+  ShellList: TCastleShellListView;
+  SelectedFileName: String;
+  SelectedURL: String;
+begin
+  if Source is TCastleShellListView then
+  begin
+    ShellList := TCastleShellListView(Source);
+    SelectedFileName := ShellList.GetPathFromItem(ShellList.Selected);
+    SelectedURL := FilenameToURISafe(SelectedFileName);
+
+    Accept := TFileFilterList.Matches(LoadScene_FileFilters, SelectedURL);
+  end else
+    Accept := false;
+end;
+
+procedure TDesignFrame.CastleControlDragDrop(Sender, Source: TObject; X, Y: Integer);
+var
+  ShellList: TCastleShellListView;
+  SelectedFileName: String;
+  SelectedURL: String;
+  Scene: TCastleScene;
+  UI: TCastleUserInterface;
+  Viewport: TCastleViewport;
+  ScenePos: TVector3;
+  RayOrigin, RayDirection: TVector3;
+  RayHit: TRayCollision;
+  Distance: Single;
+  OldPickable: Boolean;
+
+  function Is2DAsset(URL : String): Boolean;
+  var
+    MimeType: String;
+  begin
+    MimeType := URIMimeType(URL);
+    Result := LoadImage_FileFilters.Matches(URL) or
+      (MimeType = 'application/json') or
+      (MimeType = 'application/x-starling-sprite-sheet') or
+      (MimeType = 'application/x-plist') or
+      (MimeType = 'application/x-cocos2d-sprite-sheet');
+  end;
+
+begin
+  if Source is TCastleShellListView then
+  begin
+    ShellList := TCastleShellListView(Source);
+    SelectedFileName := ShellList.GetPathFromItem(ShellList.Selected);
+    SelectedURL := MaybeUseDataProtocol(FilenameToURISafe(SelectedFileName));
+
+    if not TFileFilterList.Matches(LoadScene_FileFilters, SelectedURL) then
+      Exit;
+
+    UI := FDesignerLayer.HoverUserInterface(Vector2(X, Y));
+    if not (UI is TCastleViewport) then
+      Exit;
+
+    Viewport := TCastleViewport(UI);
+    try
+      Scene := AddComponent(Viewport.Items, TCastleScene, nil) as TCastleScene;
+
+      if Is2DAsset(SelectedURL) then
+        Scene.Setup2D;
```

Comment:
In the end, I think we can just remove 

```pascal
if Is2DAsset(SelectedURL) then
  Scene.Setup2D;
```

and remove hardcoded `Is2DAsset` detection. It's my fault, I was confusing you and myself when we talked about this :) Thinking about it clearly now, this `Scene.Setup2D` (which does `RenderOptions.BlendingSort := bs2D`) just doesn't matter in this case:

- for images or sprite sheets, `RenderOptions.BlendingSort` don't matter. They only show 1 layer.

- for Spine models, `RenderOptions.BlendingSort` don't matter, as imported Spine models use `NavigationInfo` to request blendingSort = 2D anyway.

All in all, calling `Scene.Setup2D` is only really useful if you import a model from a "general" (3D or 2D) format like glTF or X3D, but you want to use it as 2D, and the model has multiple layers that use blending. I made a note about it on https://github.com/castle-engine/castle-engine/wiki/Blending . Doing it automatically, in a hardcoded subset of cases, may be more confusing than helpful.